File: ui/src/api/tables.ts
import { loggedInvoke } from '@/utils/logged-invoke';

/** A table in the floor plan with position, capacity, and status. */
export interface Table {
  id: string;
  name: string;
  capacity: number;
  pos_x: number;
  pos_y: number;
  shape: string;
  width: number;
  height: number;
  status: string;
  active_sale_id: string | null;
  section: string;
  active: boolean;
  sort_order: number;
}

/** List all tables, optionally filtered by section (session-scoped — ADR #7). */
export const listTables = (sessionToken: string, section?: string) =>
  loggedInvoke<Table[]>('list_tables_scoped', { sessionToken, section: section ?? null });

/** List tables (scoped — ADR #7). */
export const listTablesScoped = (sessionToken: string, section?: string) =>
  loggedInvoke<Table[]>('list_tables_scoped', { sessionToken, section: section ?? null });

/** Get a single table by its identifier (session-scoped — ADR #7). */
export const getTable = (sessionToken: string, id: string) =>
  loggedInvoke<Table | null>('get_table_scoped', { sessionToken, id });

/** Get a table (scoped — ADR #7). */
export const getTableScoped = (sessionToken: string, id: string) =>
  loggedInvoke<Table | null>('get_table_scoped', { sessionToken, id });

/** Create a new table in the floor plan (session-scoped — ADR #7). */
export const createTable = (sessionToken: string, table: Table) =>
  loggedInvoke<Table>('create_table_scoped', { sessionToken, table });

/** Create a table (scoped — ADR #7). */
export const createTableScoped = (sessionToken: string, table: Table) =>
  loggedInvoke<Table>('create_table_scoped', { sessionToken, table });

/** Update an existing table (session-scoped — ADR #7). */
export const updateTable = (sessionToken: string, table: Table) =>
  loggedInvoke<Table>('update_table_scoped', { sessionToken, table });

/** Update a table (scoped — ADR #7). */
export const updateTableScoped = (sessionToken: string, table: Table) =>
  loggedInvoke<Table>('update_table_scoped', { sessionToken, table });

/** Delete a table from the floor plan (session-scoped — ADR #7). */
export const deleteTable = (sessionToken: string, id: string) =>
  loggedInvoke<void>('delete_table_scoped', { sessionToken, id });

/** Delete a table (scoped — ADR #7). */
export const deleteTableScoped = (sessionToken: string, id: string) =>
  loggedInvoke<void>('delete_table_scoped', { sessionToken, id });

/** Update a table's status (e.g. free, occupied, reserved) (session-scoped — ADR #7). */
export const updateTableStatus = (sessionToken: string, id: string, status: string) =>
  loggedInvoke<Table>('update_table_status_scoped', { sessionToken, id, status });

/** Update table status (scoped — ADR #7). */
export const updateTableStatusScoped = (sessionToken: string, id: string, status: string) =>
  loggedInvoke<Table>('update_table_status_scoped', { sessionToken, id, status });

/** Assign an active sale (order) to a table (session-scoped — ADR #7). */
export const assignTableOrder = (sessionToken: string, tableId: string, saleId: string) =>
  loggedInvoke<Table>('assign_table_order_scoped', { sessionToken, tableId, saleId });

/** Assign order to table (scoped — ADR #7). */
export const assignTableOrderScoped = (sessionToken: string, tableId: string, saleId: string) =>
  loggedInvoke<Table>('assign_table_order_scoped', { sessionToken, tableId, saleId });

/** Release a table, clearing its active order assignment (session-scoped — ADR #7). */
export const releaseTable = (sessionToken: string, tableId: string) =>
  loggedInvoke<Table>('release_table_scoped', { sessionToken, tableId });

/** Release a table (scoped — ADR #7). */
export const releaseTableScoped = (sessionToken: string, tableId: string) =>
  loggedInvoke<Table>('release_table_scoped', { sessionToken, tableId });

/** List all table sections (session-scoped — ADR #7). */
export const listSections = (sessionToken: string) =>
  loggedInvoke<string[]>('list_sections_scoped', { sessionToken });

/** List sections (scoped — ADR #7). */
export const listSectionsScoped = (sessionToken: string) =>
  loggedInvoke<string[]>('list_sections_scoped', { sessionToken });

File: apps/desktop-client/src/commands/tables.rs
//! Table management Tauri commands.
//!
//! CRUD for restaurant floor-plan tables plus section listing.
//!
//! All commands have scoped variants (ADR #7) that use the session token
//! pattern. Old commands are preserved with deprecation notices.

use tauri::State;

use oz_core::Table;
use oz_core::db::Store;

use crate::commands::authz::require_permission_for_session;
use crate::error::AppError;
use crate::state::AppState;

// ── Read Commands ────────────────────────────────────────────────────

/// List tables for the store resolved from a session token. ADR #7.
#[tauri::command]
pub async fn list_tables_scoped(
    session_token: String,
    section: Option<String>,
    state: State<'_, AppState>,
) -> Result<Vec<Table>, AppError> {
    let conn = state.resolve_store(&session_token)?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    let tables = store.list_tables(section.as_deref())?;
    drop(db);
    Ok(tables)
}

/// Get a table from the store resolved from a session token. ADR #7.
#[tauri::command]
pub async fn get_table_scoped(
    session_token: String,
    id: String,
    state: State<'_, AppState>,
) -> Result<Option<Table>, AppError> {
    let conn = state.resolve_store(&session_token)?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    let table = store.get_table(&id)?;
    drop(db);
    Ok(table)
}

/// List sections for the store resolved from a session token. ADR #7.
#[tauri::command]
pub async fn list_sections_scoped(
    session_token: String,
    state: State<'_, AppState>,
) -> Result<Vec<String>, AppError> {
    let conn = state.resolve_store(&session_token)?;
    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    let sections = store.list_sections()?;
    drop(db);
    Ok(sections)
}

// ── Write Commands ───────────────────────────────────────────────────

/// Create a table in the store resolved from a session token. ADR #7.
#[tauri::command]
pub async fn create_table_scoped(
    session_token: String,
    table: Table,
    state: State<'_, AppState>,
) -> Result<Table, AppError> {
    let session = state.resolve_session(&session_token)?;
    require_permission_for_session(&state, &session, oz_core::permissions::TABLES_CREATE).await?;
    let conn = state
        .db_manager
        .open_store(&session.store_id)
        .map_err(|e| AppError::Internal(format!("opening store db: {e}")))?;

    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    let result = store.create_table(&table)?;
    drop(db);
    Ok(result)
}

/// Update a table in the store resolved from a session token. ADR #7.
#[tauri::command]
pub async fn update_table_scoped(
    session_token: String,
    table: Table,
    state: State<'_, AppState>,
) -> Result<Table, AppError> {
    let session = state.resolve_session(&session_token)?;
    require_permission_for_session(&state, &session, oz_core::permissions::TABLES_EDIT).await?;
    let conn = state
        .db_manager
        .open_store(&session.store_id)
        .map_err(|e| AppError::Internal(format!("opening store db: {e}")))?;

    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    let result = store.update_table(&table)?;
    drop(db);
    Ok(result)
}

/// Delete a table in the store resolved from a session token. ADR #7.
#[tauri::command]
pub async fn delete_table_scoped(
    session_token: String,
    id: String,
    state: State<'_, AppState>,
) -> Result<(), AppError> {
    let session = state.resolve_session(&session_token)?;
    require_permission_for_session(&state, &session, oz_core::permissions::TABLES_DELETE).await?;
    let conn = state
        .db_manager
        .open_store(&session.store_id)
        .map_err(|e| AppError::Internal(format!("opening store db: {e}")))?;

    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    store.delete_table(&id)?;
    drop(db);
    Ok(())
}

/// Update a table's status in the store resolved from a session token. ADR #7.
#[tauri::command]
pub async fn update_table_status_scoped(
    session_token: String,
    id: String,
    status: String,
    state: State<'_, AppState>,
) -> Result<Table, AppError> {
    let session = state.resolve_session(&session_token)?;
    require_permission_for_session(&state, &session, oz_core::permissions::TABLES_CLOSE).await?;
    let conn = state
        .db_manager
        .open_store(&session.store_id)
        .map_err(|e| AppError::Internal(format!("opening store db: {e}")))?;

    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    let table = store.update_table_status(&id, &status)?;
    drop(db);
    Ok(table)
}

/// Assign an order to a table in the store resolved from a session token. ADR #7.
#[tauri::command]
pub async fn assign_table_order_scoped(
    session_token: String,
    table_id: String,
    sale_id: String,
    state: State<'_, AppState>,
) -> Result<Table, AppError> {
    let session = state.resolve_session(&session_token)?;
    require_permission_for_session(&state, &session, oz_core::permissions::TABLES_ASSIGN).await?;
    let conn = state
        .db_manager
        .open_store(&session.store_id)
        .map_err(|e| AppError::Internal(format!("opening store db: {e}")))?;

    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    let table = store.assign_table_order(&table_id, &sale_id)?;
    drop(db);
    Ok(table)
}

/// Release a table in the store resolved from a session token. ADR #7.
#[tauri::command]
pub async fn release_table_scoped(
    session_token: String,
    table_id: String,
    state: State<'_, AppState>,
) -> Result<Table, AppError> {
    let session = state.resolve_session(&session_token)?;
    require_permission_for_session(&state, &session, oz_core::permissions::TABLES_CLOSE).await?;
    let conn = state
        .db_manager
        .open_store(&session.store_id)
        .map_err(|e| AppError::Internal(format!("opening store db: {e}")))?;

    let db = conn
        .lock()
        .map_err(|e| AppError::Internal(format!("store db lock: {e}")))?;
    let store = Store::new(&db);
    let table = store.release_table(&table_id)?;
    drop(db);
    Ok(table)
}

#[cfg(test)]
#[path = "tables_tests.rs"]
mod tests;
